modules/phase_field/src/materials/GBAnisotropyBase.C

```diff
@@ -0,0 +1,170 @@
+  return params;
+}
+
+GBAnisotropyBase::GBAnisotropyBase(const InputParameters & parameters) :
+    Material(parameters),
+    _mesh_dimension(_mesh.dimension()),
+    _length_scale(getParam<Real>("length_scale")),
+    _time_scale(getParam<Real>("time_scale")),
+    _M_V(getParam<Real>("molar_volume_value")),
+    _delta_sigma(getParam<Real>("delta_sigma")),
+    _delta_mob(getParam<Real>("delta_mob")),
+    _Anisotropic_GB_file_name(getParam<FileName>("Anisotropic_GB_file_name")),
+    _inclination_anisotropy(getParam<bool>("inclination_anisotropy")),
+    _T(coupledValue("T")),
+    _kappa(declareProperty<Real>("kappa_op")),
+    _gamma(declareProperty<Real>("gamma_asymm")),
+    _L(declareProperty<Real>("L")),
+    _mu(declareProperty<Real>("mu")),
+    _molar_volume(declareProperty<Real>("molar_volume")),
+    _entropy_diff(declareProperty<Real>("entropy_diff")),
+    _act_wGB(declareProperty<Real>("act_wGB")),
+    _tgrad_corr_mult(declareProperty<Real>("tgrad_corr_mult")),
+    _kb(8.617343e-5), // Boltzmann constant in eV/K
+    _JtoeV(6.24150974e18), // Joule to eV conversion
+    _mu_qp(0.0),
+    _op_num(coupledComponents("v")),
+    _vals(_op_num),
+    _grad_vals(_op_num)
+{
+  // reshape vectors
+  _sigma.resize(_op_num);
+  _mob.resize(_op_num);
+  _Q.resize(_op_num);
+  _kappa_gamma.resize(_op_num);
+  _a_g2.resize(_op_num);
+
+  for (unsigned int op = 0; op < _op_num; ++op)
+  {
+    // Initialize variables
+    _vals[op] = &coupledValue("v", op);
+    _grad_vals[op] = &coupledGradient("v", op);
+
+    _sigma[op].resize(_op_num);
+    _mob[op].resize(_op_num);
+    _Q[op].resize(_op_num);
+    _kappa_gamma[op].resize(_op_num);
+    _a_g2[op].resize(_op_num);
+  }
+
+  // Read in data from "Anisotropic_GB_file_name"
+  std::ifstream inFile(_Anisotropic_GB_file_name.c_str());
+
+  if (!inFile)
+    mooseError("Can't open GB anisotropy input file");
+
+  for (unsigned int i = 0; i < 2; ++i)
+    inFile.ignore(255, '\n'); // ignore line
+
+  Real data;
+  for (unsigned int i = 0; i < 3 * _op_num; ++i)
+  {
+    std::vector<Real> row; // create an empty row of double values
+    for (unsigned int j = 0; j < _op_num; ++j)
+    {
+      inFile >> data;
+      row.push_back(data);
+    }
+
+    if (i < _op_num)
+      _sigma[i] = row; // unit: J/m^2
+
+    else if (i < 2 * _op_num)
+      _mob[i-_op_num] = row; // unit: m^4/(J*s)
+
+    else
+      _Q[i - 2 * _op_num] = row; // unit: eV
+  }
+
+  inFile.close();
+}
+
+void
+GBAnisotropyBase::computeQpProperties()
+{
+  std::vector<std::vector<Real> > MOB = _mob;
+
+  Real sum_kappa = 0.0;
+  Real sum_gamma = 0.0;
+  Real sum_L = 0.0;
+  Real Val = 0.0;
+  Real sum_val = 0.0;
+  Real f_sigma = 1.0;
+  Real f_mob = 1.0;
+  Real gamma_value = 0.0;
+
+  for (unsigned int m = 0; m < _op_num - 1; ++m)
+  {
+    for (unsigned int n = m + 1; n < _op_num; ++n) // m<n
+    {
+      MOB[m][n] *= std::exp(-_Q[m][n] / (_kb * _T[_qp])); // Arrhenius relation
```

Comment:
drop `MOB` entirely. it does not save you anything and costs you a vector copy (plus its naming is odd).